src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,520 @@
+#if DEBUG
+        // Ensure that the bounds check that we're removing is in the fast path:
+        bool statementFound = false;
+        for (Statement* const stmt : fastpathBb->Statements())
+        {
+            if (stmt == info.stmt)
+            {
+                statementFound = true;
+
+                // Find the bndChk in the statement
+                Compiler::fgWalkResult result = comp->fgWalkTreePre(
+                    stmt->GetRootNodePointer(),
+                    [](GenTree** pTree, Compiler::fgWalkData* data) -> Compiler::fgWalkResult {
+                    return (*pTree == (GenTree*)data->pCallbackData) ? Compiler::WALK_ABORT : Compiler::WALK_CONTINUE;
+                },
+                    info.bndChk);
+                // We don't need to validate bndChkParent - optRemoveRangeCheck will do it for us
+                assert(result == Compiler::WALK_ABORT);
+                break;
+            }
+        }
+        assert(statementFound);
+#endif
+        comp->optRemoveRangeCheck(info.bndChk, info.bndChkParent, info.stmt);
+        comp->gtSetStmtInfo(info.stmt);
+        comp->fgSetStmtSeq(info.stmt);
+    }
+
+    comp->fgMorphBlockStmt(lowerBndBb, lowerBndBb->lastStmt() DEBUGARG("Morph lowerBnd"));
+    comp->fgMorphBlockStmt(upperBndBb, upperBndBb->lastStmt() DEBUGARG("Morph upperBnd"));
+    if (lowerBndBb->lastStmt() != nullptr)
+    {
+        // lowerBndBb might be converted into no-op by fgMorphBlockStmt(lowerBndBb)
+        // it happens when we emit BBJ_COND(0 >= 0) fake block (for simplicity)
+        comp->gtUpdateStmtSideEffects(lowerBndBb->lastStmt());
+    }
+    comp->gtUpdateStmtSideEffects(upperBndBb->lastStmt());
+
+    // All blocks must be in the same EH region
+    assert(BasicBlock::sameEHRegion(prevBb, lowerBndBb));
+    assert(BasicBlock::sameEHRegion(prevBb, upperBndBb));
+    assert(BasicBlock::sameEHRegion(prevBb, fastpathBb));
+    assert(BasicBlock::sameEHRegion(prevBb, fallbackBb));
+    assert(BasicBlock::sameEHRegion(prevBb, lastBb));
+
+    return fastpathBb;
+}
+
+// A visitor to record all the bounds checks in a statement in the execution order
+class BoundsChecksVisitor final : public GenTreeVisitor<BoundsChecksVisitor>
+{
+    Statement*                      m_stmt;
+    ArrayStack<BoundCheckLocation>* m_boundsChks;
+
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    BoundsChecksVisitor(Compiler* compiler, Statement* stmt, ArrayStack<BoundCheckLocation>* bndChkLocations)
+        : GenTreeVisitor(compiler)
+        , m_stmt(stmt)
+        , m_boundsChks(bndChkLocations)
+    {
+    }
+
+    fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* node = *use;
+        if (node->OperIs(GT_BOUNDS_CHECK))
+        {
+            m_boundsChks->Push(BoundCheckLocation(m_stmt, node->AsBoundsChk(), user));
+        }
+        return fgWalkResult::WALK_CONTINUE;
+    }
+};
```

Comment:
If you do this in preorder you can use `GTF_EXCEPT` to quickly locate the bounds checks in each statement.